oeps/oep-0044-bp-changelogs.rst

```diff
@@ -0,0 +1,285 @@
+
+  2020-03-21
+  ----------
+
+  Added
+  ~~~~~
+
+  * Feature x. See announcement here: https://xxx
+
+Notable Differences
+~~~~~~~~~~~~~~~~~~~
+
+The following are notable differences between the recommendations of this OEP and the `keepachangelog.com`_ recommendations:
+
+* We use GitHub Releases in certain cases, as you'll see below, for developer efficiency purposes.
+* We sometimes use automation that uses git commit messages for changelog messages. As noted above, it is acceptable and appropriate to clean these up as needed.
+
+Library Repositories
+--------------------
+
+All Open edX library repositories should contain a CHANGELOG.rst. These libraries should all adhere to `Semantic Versioning`_.
+
+Semantic Versioning
+~~~~~~~~~~~~~~~~~~~
+
+All versioned code should use a MAJOR.MINOR.PATCH format for versioning, as defined by the `Semantic Versioning v2`_ specification. Here is a copy of the most relevant part of the summary:
+
+Given a version number MAJOR.MINOR.PATCH, increment the:
+
+* *MAJOR* version when you make *incompatible API changes*,
+* *MINOR* version when you *add functionality in a backwards compatible manner*, and
+* *PATCH* version when you make *backwards compatible bug fixes*.
+
+Every published package should have a corresponding changelog entry for the new version.
+
+One of the most important rules when choosing whether to increment the MAJOR, MINOR, or PATCH version, is that backward incompatible changes always result in a new MAJOR version, and never a MINOR or PATCH version.
+
+You must increment the MAJOR version for backward incompatible changes, even if you think it is unlikely to affect anyone. You can simply note in the changelog entry why it is backward incompatible and why you think it is unlikely to affect anyone.
+
+If you find a backward incompatible change in a MINOR or PATCH version after publishing the package, please update the changelog and communicate the inconsistency. Additionally, you have the following options:
+
+* A new MINOR or PATCH version can be immediately released to revert the change, and the change can be restored with a new MAJOR version. Or,
+* If you decide that reverting will cause more harm for some reason, and you do not intend to revert, please let this be a part of your communication.
+
+The reason it is important that only MAJOR versions include backward incompatible change is because when people upgrade dependencies, some will only read changelogs for MAJOR versions because of the potential backward incompatibilities, and may not read notes for MINOR or PATCH upgrades.
+
+Frontend Libraries
+~~~~~~~~~~~~~~~~~~
+
+Like all Open edX repositories, frontend libraries should also include a CHANGELOG.rst. However, as you'll see below, this file will simply point to the GitHub Releases. Here is an example CHANGELOG.rst::
```

Comment:
> However, as you'll see below, this file will simply point to the GitHub Releases.

:disappointed: 

This is really too bad. It's great to be able to find all changes in a single file, to be easily read by humans, something that is not allowed by semantic-releases. Also, semantic-releases picks changes from git commits, which directly contradicts the principle according to which changelogs are made for humans, not machines. A changelog entry should be allowed to be longer than 50 characters, which is the standard git commit title maximum length.

Any chance we can improve on that?

Reply:
1. As noted in several places in the OEP, initial changelog entries can and should be updated as-needed, so there are no limitations.
2. Regarding a single file, I won't stop anyone from finding or writing a tool that reads the GitHub Releases and produces a file. :)

Reply:
1. To be blunt, I don't expect Open edX developers to come back and edit their commit messages. They simply don't have any incentive to do so. Case in point: https://github.com/edx/frontend-build/releases The release messages make no sense at all for those who are not already experts of this component.
2. If we publish a changelog from parsing the github releases, it will not be in sync with the git tags. Let me suggest something else: I'll write a tool that parses the changelog and will generate the github release message.

Reply:
1a: When you say "edit their commit messages", do you mean edit their changelog messages?
1b: I don't know if you are asking for different changelog messages, or fewer versions.  Assuming we agree that there would be a changelog for each version, maybe you could provide example changelog messages for the recent versions of frontend-build that you would be happy with?  Or, if you simply want fewer versions, how do you propose we adjust our process to make that happen?
2. I am all for new tools that help improve all this.  Just keep in mind that the current tool is maintained, and any new tools would need a maintainer.  Also, the tool would need to perform additional tasks, including keeping the two in sync over time if one changes, and the process for publishing to npm would change, etc.  This is all possible, but this is certainly a larger project.

Reply:
> 1a: When you say "edit their commit messages", do you mean edit their changelog messages?

Yes. If I understand correctly, the "semantic-release" utility uses the commit title as a changelog entry, so in the proposed approach commit titles are changelog messages, and vice versa.

> 1b: I don't know if you are asking for different changelog messages, or fewer versions. Assuming we agree that there would be a changelog for each version, maybe you could provide example changelog messages for the recent versions of frontend-build that you would be happy with? Or, if you simply want fewer versions, how do you propose we adjust our process to make that happen?

I'm perfectly happy with many new versions a day. But the changelog messages should be targeted towards users. As such, changelog entries such as "fix tests" or "internal refactor of FooBar class" are totally useless to me. When there are only such changes in a release, I'd be happy to read "no functional change" for a release. When there are indeed changes, they could be phrased as such:

- "Fix 500 error during logout" (and not: "Fix SyntaxError in UserLogout class")
- "Extend the public API with a `set_foobarizer` method: this is particularly useful for application developers trying to foobarizing their users. See [the docs](http://thedocs.com/set_foobarizer). (and not: "Add `set_foobarizer` method to `api.Foo`)
- [Breaking change] Get rid of the `dry_run` parameter in the public `foobarize` API method. This parameter is deprecated in favour of the `no_apply` parameter. See [the docs](http://thedocs.com/no_apply). ("Rename `dry_run` parameter)

2. Hey, you are the one who suggested we create a new tool ;) There is no need to keep a CHANGELOG.md file and the Github releases in sync. Having a changelog would already be a major improvement over the current process. Older changelog entries are typically not updated very often. Also, the Changelog is versioned, so it's usually regarded as a better source of authority than to Github releases. So I think it's ok if we sometimes make changes to older entries of a changelog file and not update the corresponding releases.

Reply:
@regisb @nedbat: Lots of great points that have me thinking about a bunch of potential updates.  I will put up changes in the next day or so, or ask more questions.  Thanks.

Reply:
@regisb:
1. You all changed my tune and I am recommending against Github Releases.
2. I think the frontend library used can write to a changelog instead, but someone needs to trial that.
3. I lifted a bunch of your examples and added them to the OEP.  Happy to add you as co-author if you want.  That's up to you.
Please mark as resolved when and if you are satisfied with this thread.  Thanks.

Reply:
:+1: